ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,576 @@
+		if err := <-errc; err != nil {
+			d.log.Error("Metrics collection failed", "err", err)
+		}
+		d.quitChan = nil
+	}
+	return d.db.Close()
+}
+
+// Has retrieves if a key is present in the key-value store.
+func (d *Database) Has(key []byte) (bool, error) {
+	_, closer, err := d.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	closer.Close()
+	return true, nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (d *Database) Get(key []byte) ([]byte, error) {
+	dat, closer, err := d.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]byte, len(dat))
+	copy(ret, dat)
+	closer.Close()
+	return ret, nil
+}
+
+// Put inserts the given value into the key-value store.
+func (d *Database) Put(key []byte, value []byte) error {
+	return d.db.Set(key, value, pebble.NoSync)
+}
+
+// Delete removes the key from the key-value store.
+func (d *Database) Delete(key []byte) error {
+	return d.db.Delete(key, nil)
+}
+
+// NewBatch creates a write-only key-value store that buffers changes to its host
+// database until a final write is called.
+func (d *Database) NewBatch() ethdb.Batch {
+	return &batch{
+		b: d.db.NewBatch(),
+	}
+}
+
+// NewBatchWithSize creates a write-only database batch with pre-allocated buffer.
+// TODO can't do this with pebble.  Batches are allocated in a pool so maybe this doesn't matter?
+func (d *Database) NewBatchWithSize(_ int) ethdb.Batch {
+	return &batch{
+		b: d.db.NewBatch(),
+	}
+}
+
+// snapshot wraps a pebble snapshot for implementing the Snapshot interface.
+type snapshot struct {
+	db *pebble.Snapshot
+}
+
+// Has retrieves if a key is present in the snapshot backing by a key-value
+// data store.
+func (snap *snapshot) Has(key []byte) (bool, error) {
+	_, closer, err := snap.db.Get(key)
+	if err != nil {
+		if err != pebble.ErrNotFound {
+			return false, err
+		} else {
+			return false, nil
+		}
+	}
+	closer.Close()
+	return true, nil
+}
+
+// Get retrieves the given key if it's present in the snapshot backing by
+// key-value data store.
+func (snap *snapshot) Get(key []byte) ([]byte, error) {
+	dat, closer, err := snap.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]byte, len(dat))
+	copy(ret, dat)
+	closer.Close()
+	return ret, nil
+}
+
+// Release releases associated resources. Release should always succeed and can
+// be called multiple times without causing error.
+func (snap *snapshot) Release() {
+	snap.db.Close()
+}
+
+// NewSnapshot creates a database snapshot based on the current state.
+// The created snapshot will not be affected by all following mutations
+// happened on the database.
+// Note don't forget to release the snapshot once it's used up, otherwise
+// the stale data will never be cleaned up by the underlying compactor.
+func (d *Database) NewSnapshot() (ethdb.Snapshot, error) {
+	snap := d.db.NewSnapshot()
+	return &snapshot{db: snap}, nil
+}
```

Comment:
Please move this up right below the `type snapshot struct` declaration. It's good practice to have the constructor as the first method after the type